aptos-move/aptos-vm-types/src/resolver.rs

```diff
@@ -1,21 +1,142 @@
+use aptos_aggregator::resolver::TAggregatorView;
+use aptos_state_view::StateViewId;
+use aptos_types::state_store::{
+    state_key::StateKey,
+    state_storage_usage::StateStorageUsage,
+    state_value::{StateValue, StateValueMetadataKind},
+};
 use bytes::Bytes;
-use move_core_types::language_storage::StructTag;
+use move_core_types::{language_storage::StructTag, value::MoveTypeLayout};
+
+/// Allows to query resources from the state.
+pub trait TResourceView {
+    type Key;
+    type Layout;
+
+    /// Returns
+    ///   -  Ok(None)         if the resource is not in storage,
+    ///   -  Ok(Some(...))    if the resource exists in storage,
+    ///   -  Err(...)         otherwise (e.g. storage error).
+    fn get_resource_state_value(
+        &self,
+        key: &Self::Key,
+        maybe_layout: Option<&Self::Layout>,
+    ) -> anyhow::Result<Option<StateValue>>;
+
+    fn get_resource_bytes(
+        &self,
+        key: &Self::Key,
+        maybe_layout: Option<&Self::Layout>,
+    ) -> anyhow::Result<Option<Bytes>> {
+        let maybe_state_value = self.get_resource_state_value(key, maybe_layout)?;
+        Ok(maybe_state_value.map(|state_value| state_value.bytes().clone()))
+    }
+
+    fn get_resource_state_value_metadata(
+        &self,
+        key: &Self::Key,
+    ) -> anyhow::Result<Option<StateValueMetadataKind>> {
+        // For metadata, layouts are not important.
+        let maybe_state_value = self.get_resource_state_value(key, None)?;
+        Ok(maybe_state_value.map(StateValue::into_metadata))
+    }
+
+    fn resource_exists(&self, key: &Self::Key) -> anyhow::Result<bool> {
+        // For existence, layouts are not important.
+        self.get_resource_state_value(key, None)
+            .map(|maybe_state_value| maybe_state_value.is_some())
+    }
+}
+
+pub trait ResourceResolver: TResourceView<Key = StateKey, Layout = MoveTypeLayout> {}
+
+impl<T: TResourceView<Key = StateKey, Layout = MoveTypeLayout>> ResourceResolver for T {}
+
+/// Allows to query modules from the state.
+pub trait TModuleView {
+    type Key;
+
+    /// Returns
+    ///   -  Ok(None)         if the module is not in storage,
+    ///   -  Ok(Some(...))    if the module exists in storage,
+    ///   -  Err(...)         otherwise (e.g. storage error).
+    fn get_module_state_value(&self, key: &Self::Key) -> anyhow::Result<Option<StateValue>>;
+
+    fn get_module_bytes(&self, key: &Self::Key) -> anyhow::Result<Option<Bytes>> {
+        let maybe_state_value = self.get_module_state_value(key)?;
+        Ok(maybe_state_value.map(|state_value| state_value.bytes().clone()))
+    }
+
+    fn get_module_state_value_metadata(
+        &self,
+        key: &Self::Key,
+    ) -> anyhow::Result<Option<StateValueMetadataKind>> {
+        let maybe_state_value = self.get_module_state_value(key)?;
+        Ok(maybe_state_value.map(StateValue::into_metadata))
+    }
+
+    fn module_exists(&self, key: &Self::Key) -> anyhow::Result<bool> {
```

Comment:
@zekun000 @davidiw I remember there was an use-case where we needed a simple existence check but didn't have it (I thought it was for a module for metadata maybe? not sure - hope you can recall). Do you remember where it is and if it still exists? We can then connect to this already (and then the default implementation will be changed to a proper parallel execution implementation - with easier validation essentially removing R/W conflict)